mayastor/src/rebuild.rs

```diff
@@ -385,3 +450,78 @@ impl RebuildOperations for RebuildJob {
         }
     }
 }
+
+impl RebuildJob {
+    fn start_all_tasks(&mut self) {
+        assert_eq!(self.tasks.active, 0, "{} active tasks", self.tasks.active);
+
+        for n in 0 .. self.tasks.total {
+            self.next = match self.send_segment_task(n) {
+                Some(next) => next,
+                None => break, /* we've already got enough tasks to rebuild
+                                * the bdev */
+            };
+
+            self.tasks.active += 1;
+        }
+    }
+
+    fn start_task_by_id(&mut self, id: u64) {
+        match self.send_segment_task(id) {
+            Some(next) => {
+                self.tasks.active += 1;
+                self.next = next;
+            }
+            None => {
+                if self.tasks.active == 0 {
+                    self.state = RebuildState::Completed;
+                }
+            }
+        };
+    }
+
+    async fn await_one_task(&mut self) -> Option<TaskResult> {
+        self.tasks.channel.1.next().await.map(|f| {
+            self.tasks.active -= 1;
+            f
+        })
+    }
+
+    async fn await_all_tasks(&mut self) {
+        while self.await_one_task().await.is_some() {
+            self.tasks.active -= 1;
+            if self.tasks.active == 0 {
+                break;
+            }
+        }
+    }
+
+    /// Sends one segment worth of data in a reactor future and notifies the
+    /// management channel. Returns the next segment offset to rebuild, if any
+    fn send_segment_task(&self, id: u64) -> Option<u64> {
+        if self.next >= self.end {
+            None
+        } else {
+            let blk = self.next;
+            let next =
+                std::cmp::min(self.next + self.segment_size_blks, self.end);
```

Comment:
Given that the test on line 502 is for ">=" we could just set next to "self.next + self.segment_size_blks" rather than optionally setting it to self.end (although this is more correct!)

Reply:
I think I prefer keeping `end` rather than having to remember that next could be pointing beyond the end of the disk. Eg, if we want to calculate progress

Reply:
That's fine